api/xds_protocol.rst

```diff
@@ -613,6 +613,43 @@ adding/removing/updating clusters. On the other hand, routes are not
 warmed, i.e., the management plane must ensure that clusters referenced
 by a route are in place, before pushing the updates for a route.
 
+.. _xds_protocol_TTL:
+
+TTL
+~~~
+
+In the event that the management server becomes unreachable, the last known configuration received
+by Envoy will persist until the connection is reestablished. For some services, this may not be
+desirable. For example, in the case of a fault injection service, a management server crash at the
+wrong time may leave Envoy in an undesirable state. The TTL setting allows Envoy to remove a set of
+resources after a specified period of time if contact with the management server is lost. This can
+be used, for example, to terminate a fault injection test when the management server can no longer
+be reached.
+
+SotW TTL
+^^^^^^^^
+
+A TTL field in the :ref:`DiscoveryResponse <envoy_api_msg_DiscoveryResponse>` specifies the TTL for
+all resources contained in the response. A timer is started when the response is received. When the
+timer expires, all resources retrieved from the management server will be removed.
+
+The TTL timer can be updated by sending a specially formed 
+:ref:`DiscoveryResponse <envoy_api_msg_DiscoveryResponse>` with the same version as the previously
+applied response. The list of resources included in the response is empty and the TTL field is
```

Comment:
For all resource types except for LDS and CDS, each individual resource can be sent in its own response, but the version will change for each response on the stream. So this would require the client to remember which set of resources were present in each version that the server sent it for the entire history of the stream. I don't think this is feasible.

Reply:
The proposal is that when the TTL expires for a SotW response, all resources associated with that subscriber will be removed.

Reply:
This means that the TTL semantics for SotW are significantly different than those for incremental.

If we could use the `Resource` proto for the SotW API, that seems like a much cleaner and more consistent way of doing this.

Reply:
I tend to agree but using Resource for SotW will break wire compatibility and require a major version bump. This would take longer than my project can likely wait.

Reply:
Hmm, so the idea here is that for SotW, the TTL will apply to all resources of a specified type, not just those that happen to be in the response that set the TTL?

For example, for EDS, let's say that at time t=0, the client receives a response containing resource A that does *not* set the TTL, and then at time t=1m, it receives another response containing resource B that *does* set the TTL to 5m.  Does that mean that the TTL in the second response would be used for both A and B?  I'm guessing both.  Would the expiration time for A be t=6m, or would it be t=5m (retroactive from when A was received)?

I think the exact semantics of this need to be clearly documented.

Reply:
Also, I can see this approach being useful for things like per-filter configs, but it seems like it would really limit the usefulness of the LEDS use-case, because there we would presumably want only a subset of the endpoints to have TTLs.  If they *all* have TTLs and then the client loses contact with the management server, then the client basically stops working when the TTL expires.

Reply:
I think I'm missing something here from my own understanding. In SoTW, all resources need to be sent in each response. In your example, the first response is A, the second response is A,B, and a third response of A would remove B. Right? In this case it seems fine to have the TTL associated with the entire SoTW state? As a user, this actually makes the most sense to me logically.

Reply:
@mattklein123 The requirement that every resource must be included in every response is true only for LDS and CDS.  For every other resource type, each resource can be sent in its own response.  For details, see:

https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#grouping-resources-into-responses

Reply:
I don't think that text is correct, unless I am tremendously confused on how things work currently. cc @htuch to help clarify.

Reply:
@mattklein123 yeah, take a look at https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/eds.cc#L151. Basically, each EDS subscription is watching a single resource, there is no global coordination. So, if the ClusterLoadAssignments we want are {X, Y, Z}, the management server is free to send  {X}, wait a while, then send {Y}, etc. We only delete EDS subscriptions when the corresponding Cluster goes away.

This is one of the quirks we inherited from the REST days of v1 I think. It means that we have two classes of SoTW, LDS/CDS (true SoTW) and EDS/RDS (quasi delta, addition only).

Reply:
Right, but this is still per stream, right? Meaning, there would still be a TTL per "SoTW", even if the "SoTW" is a lot of independent streams? It's not optimal but I still think it makes sense.

Reply:
In ADS, there is just one stream which all EDS updates share.

Reply:
@mattklein123 This is one of many strange semantics in the SotW protocol that we ran into while adding xDS support to gRPC.  I would really like to ultimately move everything toward the incremental protocol, since it is much more consistent in its behavior.  Note that the current UDPA-TP draft doesn't have a SotW protocol variant.

In any case, getting back to the TTL feature, if we are saying that we will need per-resource timers for the incremental protocol anyway, I would strongly prefer to have the same capability in the SotW protocol; I think having dramatically different semantics between the two will cause confusion and make the implementation more difficult.  I think the ideal way to solve this problem would be to use the `Resource` message in the SotW protocol, as discussed above.  That would allow setting TTLs on a per-resource basis, and it would also provide a clear way to refresh the TTL for a given resource without resending the whole resource and without abusing the version.

Reply:
OK I didn't realize the ADS issue. Agreed this is inconsistent and confusing. Whether we continue to support SoTW in the future is a different/larger issue that I suggest we tackle elsewhere. Incremental is an order of magnitude more complicated to implement for control planes. Perhaps over time this is less of an issue as more people move to vendor/OSS stock control plane solutions.

I was just talking to @wgallagher offline and one other option would be to just support TTL on incremental and not support it on SoTW at all. This would be fine with me also if that is the consensus on how to move forward.

Reply:
I would also be fine with supporting TTLs only in the incremental protocol.  That might be a good carrot to offer people to migrate to the incremental protocol. :)

Reply:
The downside for us is that the go-control-plane does not yet support incremental so there may be considerable effort involved in getting that to work.